File: src/main/java/io/github/divyeshmangla/timetable/parser/DaySlotsFactory.java
package io.github.divyeshmangla.timetable.parser;

import io.github.divyeshmangla.timetable.model.Day;
import io.github.divyeshmangla.timetable.model.TimeSlot;
import org.apache.poi.ss.usermodel.Cell;

import java.util.ArrayList;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;

class DaySlotsFactory {
    private final List<DaySlots> result = new ArrayList<>();
    private final Day[] days = Day.values();
    private int currentDayIndex = 0;
    private Map<TimeSlot, Cell> currentDaySlots = new EnumMap<>(TimeSlot.class);

    void processSlot(Integer slotNumber, Cell cell) {
        if (isDayBoundary(slotNumber)) {
            finalizeCurrentDay();
        }

        try {
            TimeSlot slot = TimeSlot.fromNumber(slotNumber);
            currentDaySlots.put(slot, cell);
        } catch (IllegalArgumentException e) {
            // ignored
        }
    }

    private boolean isDayBoundary(Integer slotNumber) {
        return slotNumber == 1 && !currentDaySlots.isEmpty();
    }

    private void finalizeCurrentDay() {
        if (currentDayIndex >= days.length) {
            return;
        }

        result.add(new DaySlots(days[currentDayIndex], currentDaySlots));
        currentDayIndex++;
        currentDaySlots = new EnumMap<>(TimeSlot.class);
    }

    boolean isComplete() {
        return currentDayIndex >= days.length;
    }

    List<DaySlots> build() {
        if (!currentDaySlots.isEmpty() && currentDayIndex < days.length) {
            result.add(new DaySlots(days[currentDayIndex], currentDaySlots));
        }
        return result;
    }
}

File: internal/parser/day_slots_factory.go
package parser

import "github.com/DivyeshMangla/tiet-timetable/internal/model"

type daySlotsFactory struct {
	result          []DaySlots
	days            []model.Day
	currentDayIndex int
	currentDaySlots map[model.TimeSlot]CellLocation
}

func newDaySlotsFactory() *daySlotsFactory {
	return &daySlotsFactory{
		result:          make([]DaySlots, 0),
		days:            []model.Day{model.MON, model.TUE, model.WED, model.THU, model.FRI, model.SAT},
		currentDayIndex: 0,
		currentDaySlots: make(map[model.TimeSlot]CellLocation),
	}
}

func (f *daySlotsFactory) processSlot(slotNumber int, cell CellLocation) {
	if f.isDayBoundary(slotNumber) {
		f.finalizeCurrentDay()
	}

	slot, err := model.FromNumber(slotNumber)
	if err != nil {
		return
	}

	f.currentDaySlots[slot] = cell
}

func (f *daySlotsFactory) isDayBoundary(slotNumber int) bool {
	return slotNumber == 1 && len(f.currentDaySlots) > 0
}

func (f *daySlotsFactory) finalizeCurrentDay() {
	if f.currentDayIndex >= len(f.days) {
		return
	}

	f.result = append(f.result, DaySlots{
		Day:   f.days[f.currentDayIndex],
		Slots: f.currentDaySlots,
	})

	f.currentDayIndex++
	f.currentDaySlots = make(map[model.TimeSlot]CellLocation)
}

func (f *daySlotsFactory) isComplete() bool {
	return f.currentDayIndex >= len(f.days)
}

func (f *daySlotsFactory) build() []DaySlots {
	if len(f.currentDaySlots) > 0 && f.currentDayIndex < len(f.days) {
		f.result = append(f.result, DaySlots{
			Day:   f.days[f.currentDayIndex],
			Slots: f.currentDaySlots,
		})
	}

	return f.result
}
